src/Storages/MergeTree/MergeTreeIndexText.cpp

```diff
@@ -450,109 +401,199 @@ void MergeTreeIndexGranuleText::analyzeDictionary(MergeTreeIndexReaderStream & s
     }
 
     auto * data_buffer = stream.getDataBuffer();
-    auto * compressed_buffer = stream.getCompressedDataBuffer();
 
     /// Either retrieves a dictionary block from cache or from disk when cache is disabled.
     const auto get_dictionary_block = [&](size_t block_id)
     {
         const auto load_dictionary_block = [&] -> TextIndexDictionaryBlockCacheEntryPtr
         {
-            UInt64 offset_in_file = header->sparseIndex().getOffsetInFile(block_id);
-            compressed_buffer->seek(offset_in_file, 0);
+            UInt64 offset_in_file = sparse_index->getOffsetInFile(block_id);
+            stream.seekToMark({offset_in_file, 0});
             return std::make_shared<TextIndexDictionaryBlockCacheEntry>(deserializeDictionaryBlock(*data_buffer));
         };
 
-        if (condition_text.useDictionaryBlockCache())
-            return condition_text.dictionaryBlockCache()->getOrSet(
-                TextIndexDictionaryBlockCache::hash(state.part.getDataPartStorage().getFullPath(), state.index.getFileName(), state.index_mark, block_id),
-                load_dictionary_block);
-
-        return load_dictionary_block();
+        auto hash = TextIndexDictionaryBlockCache::hash(state.part.getDataPartStorage().getFullPath(), state.index.getFileName(), block_id);
+        return condition_text.dictionaryBlockCache()->getOrSet(hash, load_dictionary_block);
     };
 
     for (const auto & [block_idx, tokens] : block_to_tokens)
     {
         const auto dictionary_block = get_dictionary_block(block_idx);
+
         for (const auto & token : tokens)
         {
-            auto it = remaining_tokens.find(token);
-            chassert(it != remaining_tokens.end());
             auto * token_info = dictionary_block->getTokenInfo(token);
 
             if (token_info)
             {
-                ProfileEvents::increment(ProfileEvents::TextIndexBloomFilterTruePositives);
-                it->second = *token_info;
+                remaining_tokens.emplace(token, *token_info);
```

Comment:
Maybe I miss something, but I wonder in l. 425, when `getTokenInfo` can return nullptr. I guess we can assert != nullptr.

EDIT: Oh, the searched token may simply not be found in the dictionary. Hmm, maybe still add a comment in l. 429: `/// Search token found in the dictionary`